File: Macaroni/Macaroni/Source/Macaroni/Model/ContextTests.lua
require "Macaroni.Model.Context";
--require "Macaroni.Model.Namespace";
require "Macaroni.Model.Node";

local Context = Macaroni.Model.Context;
local Namespace = Macaroni.Model.Namespace;
local Node = Macaroni.Model.Node;

Test.register(
{	
name = "Context Tests",    
tests={	       
    ["CreateRoot Test"] = function(this)
        this.context = Context.New("{ROOT}");        
        Test.assert("{ROOT}", tostring(this.context.Root));        
    end,   
    
    ["ToString returns a string of some kind."] = function(this)
        local lContext = Context.New("{ROOT}");        
        Test.assert("Context[references:3,Root:{ROOT}]", tostring(lContext));        
    end,
    
    
    ["ParseComplexName will morph unknown Nodes into Namespaces."] = function(this)
        local context = Context.New("{ROOT}", "{WILDCARD}");
        local d = context.Root:FindOrCreate("A::B::C::D");
        Test.assert("A::B::C::D", d.FullName);       
        local d_c = d.Node;
        Test.assert("A::B::C", d_c.FullName);
        local d_b = d_c.Node;
        Test.assert("A::B", d_b.FullName);
        local d_a = d_b.Node;
        Test.assert("A", d_a.FullName);       
        Test.assert(1, #(context.Root.Children));
       
        -- Must reuse A, and morph it into an Namespace...
        -- I.. am.. morphing...
        local a = context.Root:FindOrCreate("A");
        Test.assert("A", a.FullName);
        Test.assert(a, d_a);
        Test.assert(1, #(context.Root.Children));
    end,
        
    --[[
    ["Existing namespace instances are used when adding to collection."] = function(this)        
        local context = Context.New("{r}", "*");
        Node.
context.RootNamespace.
        
		doc:Read("namespace A::B::C::D { } ");
		doc:Read("namespace A{}");
		Test.assert(2, #(doc.Namespaces));
		local a = doc.Namespaces[2];
		local d = doc.Namespaces[1];
		Test.assert("A::B::C::D", d.FullName);
		Test.assert("A", a.FullName);		
		local c = d.Parent;
		local b = d.Parent.Parent;		
		local aReference2 = d.Parent.Parent.Parent;
		Test.assert(a, aReference2);
    end,]]--
    
    ["Root and wildcard namespaces created properly"] = function(this)
        context = Context.New("{ROOT}");
        root = context.Root;
        Test.assert("{ROOT}", tostring(root));                
    end,    
}

} -- End of Test table.
); -- End of register call

File: Macaroni/Macaroni/Source/Macaroni/Model/ContextTests.cpp
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE ContextTests
#include <boost/test/unit_test.hpp>

#include "Context.h"
#include "Node.h"

using Macaroni::Model::Context;
using Macaroni::Model::ContextPtr;
using Macaroni::Model::Node;
using Macaroni::Model::NodePtr;

BOOST_AUTO_TEST_SUITE(ContextSuite)

BOOST_AUTO_TEST_CASE(CreateRoot_Test)
{	
	ContextPtr context(new Context("%ROOT%"));
	
	BOOST_CHECK_EQUAL(1, context->GetReferenceCount());	

	NodePtr root = context->GetRoot();
	BOOST_CHECK( !!root );
	std::string name = root->GetName();
	BOOST_CHECK_EQUAL(root->GetName(), "%ROOT%");	
}

BOOST_AUTO_TEST_CASE(ParseComplexName_will_morph_unknown_Nodes_into_Namespaces)
{
	ContextPtr context(new Context("{ROOT}"));
    NodePtr d = context->GetRoot()->FindOrCreate("A::B::C::D");
    BOOST_CHECK_EQUAL(d->GetFullName(), "A::B::C::D");       
    NodePtr d_c = d->GetNode();
	BOOST_CHECK_EQUAL(d_c->GetFullName(), "A::B::C");
    NodePtr d_b = d_c->GetNode();
    BOOST_CHECK_EQUAL(d_b->GetFullName(), "A::B");
    local d_a = d_b.Node;
    Test.assert("A", d_a.FullName);       
    Test.assert(1, #(context.Root.Children));
   
    -- Must reuse A, and morph it into an Namespace...
    -- I.. am.. morphing...
    local a = context.Root:FindOrCreate("A");
    Test.assert("A", a.FullName);
    Test.assert(a, d_a);
    Test.assert(1, #(context.Root.Children));
}

BOOST_AUTO_TEST_SUITE_END()
